### PR TITLE
chunked cross entropy loss to reduce peak memory

### PR DESCRIPTION
**Pull request by xuanzhang816**
Noticed that in the memory snapshot when running the Llama3-8B models, the cross entropy loss is causing huge spikes in peak memory: 
image

Replace the cross entropy loss function with a chunked version can help reduce peak memory by 7GiB without affecting training throughput. This is inspired by the torchtune [implementation](https://github.com/pytorch/torchtune/blob/c3703482bde72e572b535d3f7c43c81e94164ebc/torchtune/modules/loss/ce_chunked_output_loss.py#L13).

**Baseline run**
```
...
[rank0]:[titan] 2025-03-20 08:19:11,539 - root - INFO - step: 30  loss:  7.7158  memory: 49.58GiB(52.20%)  tps: 5,966  tflops: 345.53  mfu: 34.94%
[rank0]:[titan] 2025-03-20 08:19:25,291 - root - INFO - step: 40  loss:  7.3677  memory: 49.58GiB(52.20%)  tps: 5,957  tflops: 345.01  mfu: 34.88%
[rank0]:[titan] 2025-03-20 08:19:39,040 - root - INFO - step: 50  loss:  7.0890  memory: 49.58GiB(52.20%)  tps: 5,959  tflops: 345.10  mfu: 34.89%
[rank0]:[titan] 2025-03-20 08:19:52,794 - root - INFO - step: 60  loss:  6.8336  memory: 49.58GiB(52.20%)  tps: 5,957  tflops: 344.99  mfu: 34.88%
...
```

**With the changes**
```
...
[rank0]:[titan] 2025-03-20 08:29:30,780 - root - INFO - step: 30  loss:  7.7423  memory: 42.26GiB(44.49%)  tps: 5,957  tflops: 345.01  mfu: 34.89%
[rank0]:[titan] 2025-03-20 08:29:44,554 - root - INFO - step: 40  loss:  7.3368  memory: 42.26GiB(44.49%)  tps: 5,948  tflops: 344.49  mfu: 34.83%
[rank0]:[titan] 2025-03-20 08:29:58,318 - root - INFO - step: 50  loss:  7.0574  memory: 42.26GiB(44.49%)  tps: 5,953  tflops: 344.74  mfu: 34.86%
[rank0]:[titan] 2025-03-20 08:30:12,080 - root - INFO - step: 60  loss:  6.8472  memory: 42.26GiB(44.49%)  tps: 5,953  tflops: 344.79  mfu: 34.86%
...
```

**cc**: @lessw2020 @weifengpy @wconstab @yf225

### CODE REVIEW

**Review by tianyu-l — COMMENTED** [review]
Makes sense to me. Could you verify the following?
1. verify you get the same loss curves after fixing `training.seed` and setting `training.deterministic=True` before vs. after.
2. turn on TP (by default with Loss Parallel) to see if the results are still correct. For this one you need to fix the DP degree, e.g. 1st job DP2, 2nd job DP2 TP4.

**Comment by xingchensong** [comment]
Hi @xuanzhang816 , nice work ! This seems like cp, i.e. `num_output_chunks == cp_degree`, whats the difference between cp and chunked ce? can we use both of them to further reduce peak memory?

**Comment by xingchensong** [comment]
> Makes sense to me. Could you verify the following?
> 
> 1. verify you get the same loss curves after fixing `training.seed` and setting `training.deterministic=True` before vs. after.
> 2. turn on TP (by default with Loss Parallel) to see if the results are still correct. For this one you need to fix the DP degree, e.g. 1st job DP2, 2nd job DP2 TP4.

I'm also curious about whether it works when both TP and **loss parallel** are enabled.

**Comment by awgu** [comment]
Chunked cross entropy loss is not as important when you have larger TP and/or are using PP with smaller microbatch size. I wonder if it makes sense to make this configurable.

**Comment by tianyu-l** [comment]
> Chunked cross entropy loss is not as important when you have larger TP and/or are using PP with smaller microbatch size. I wonder if it makes sense to make this configurable.

I think this is the right thing to do. Let's make it configurable, necessitating switching `loss_fn` to `build_loss_fn` in `TrainSpec` and add a str config for the loss. If it sounds non-trivial, I can do it in a follow-up PR -- in this PR let's add this loss, but not making it the default.

WDYT?

**Comment by xuanzhang816** [comment]
> > Chunked cross entropy loss is not as important when you have larger TP and/or are using PP with smaller microbatch size. I wonder if it makes sense to make this configurable.
> 
> I think this is the right thing to do. Let's make it configurable, necessitating switching `loss_fn` to `build_loss_fn` in `TrainSpec` and add a str config for the loss. If it sounds non-trivial, I can do it in a follow-up PR -- in this PR let's add this loss, but not making it the default.
> 
> WDYT?

Sounds good! I will only land the chunked loss function in this PR, pending the two checks you asked on loss curves.

**Comment by xuanzhang816** [comment]
> Makes sense to me. Could you verify the following?
> 
> 1. verify you get the same loss curves after fixing `training.seed` and setting `training.deterministic=True` before vs. after.
> 2. turn on TP (by default with Loss Parallel) to see if the results are still correct. For this one you need to fix the DP degree, e.g. 1st job DP2, 2nd job DP2 TP4.

Checked these and they are correct. Analysis can be found in Bento Notebook -- [N6820684](https://www.internalfb.com/intern/anp/view/?id=6820684)

**1. compare cross_entroy_loss with chunked_cross_entropy_loss**
Screenshot 2025-03-24 at 3 41 51 PM

**2. make sure chunked_cross_entropy_loss works with TP**
Screenshot 2025-03-24 at 3 41 46 PM

**Review by tianyu-l — APPROVED** [review]
lgtm!
In another PR, let's have a builder function to make this option customizable.
